RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -0,0 +1,914 @@
+                     getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::EcalSize)) /
+                (getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::EcalSize) - 1);
+          } else {
+            getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::EcalEnergyStdDev) = 0.;
+          }
+        }
+      }
+
+      // caorechit_HBHE
+      for (const auto& caloRecHit_hbhe : *caloRecHits.hbhe) {
+        if (caloRecHit_hbhe.energy() <= 0)
+          continue;
+        auto& position = caloRecHits.Geometry->getGeometry(caloRecHit_hbhe.id())->getPosition();
+        float hbheCalEn = caloRecHit_hbhe.energy();
+        float hbheCalChi2 = caloRecHit_hbhe.chi2();
+        if (reco::deltaR2(position, allTaus[tau_idx]->polarP4()) < dR2_max) {
+          const auto [deta, dphi, eta_idx, phi_idx] = getEtaPhiIndices(position, allTaus[tau_idx]->polarP4());
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalEnergySum) += hbheCalEn;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalEnergyStdDev) += hbheCalEn * hbheCalEn;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalSize) += 1.;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalDeltaEta) += deta * hbheCalEn;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalDeltaPhi) += dphi * hbheCalEn;
+          if (hbheCalChi2 >= 0) {
+            getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalChi2) += hbheCalChi2 * hbheCalEn;
+            getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalEnergySumForPositiveChi2) += hbheCalEn;
+            getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalSizeForPositiveChi2) += 1.;
+          }
+        }
+      }
+
+      // caorechit_HO
+      for (const auto& caloRecHit_ho : *caloRecHits.ho) {
+        if (caloRecHit_ho.energy() <= 0)
+          continue;
+        auto& position = caloRecHits.Geometry->getGeometry(caloRecHit_ho.id())->getPosition();
+        float hoCalEn = caloRecHit_ho.energy();
+        if (reco::deltaR2(position, allTaus[tau_idx]->polarP4()) < dR2_max) {
+          const auto [deta, dphi, eta_idx, phi_idx] = getEtaPhiIndices(position, allTaus[tau_idx]->polarP4());
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalEnergySum) += hoCalEn;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalEnergyStdDev) += hoCalEn * hoCalEn;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalSize) += 1.;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalDeltaEta) += deta * hoCalEn;
+          getCell(tau_idx, phi_idx, eta_idx, L2TauTagNNv1::NNInputs::HcalDeltaPhi) += dphi * hoCalEn;
+        }
+      }
+
+      // normalize to sum and define stdDev
+      for (int eta_idx = 0; eta_idx < nCellEta; eta_idx++) {
```

Comment:
in `FillCaloRecHits` you have many loops, so in order to not pass `phi_idx` and `eta_idx` when calling `getCell`, you can have them defined outside.